### PR TITLE
Add claims to principal

### PR DESCRIPTION
**Pull request by louischan-oursky**
connect #969

### CODE REVIEW

File: pkg/auth/dependency/principal/customtoken/provider.go [hunk 1]
```diff
@@ -198,57 +230,66 @@ func (p providerImpl) GetPrincipalByID(principalID string) (principal.Principal,
...
+	builder := p.sqlBuilder.Select(
+		"p.id",
+		"p.user_id",
+		"ct.token_principal_id",
+		"ct.raw_profile",
+		"ct.claims",
+	).
+		From(fmt.Sprintf("%s AS p", p.sqlBuilder.FullTableName("principal"))).
+		Join(fmt.Sprintf("%s AS ct ON p.id = ct.principal_id", p.sqlBuilder.FullTableName("provider_custom_token"))).
+		Where(fmt.Sprintf(`(ct.claims #>> '{%s}') = ?`, claimName), claimValue)
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/dependency/principal/customtoken/provider.go]: Let me have a try

File: pkg/auth/dependency/principal/customtoken/mock_provider.go [hunk 2]
```diff
@@ -73,17 +72,45 @@ func (p MockProvider) GetPrincipalByTokenPrincipalID(tokenPrincipalID string) (*
...
-	attrs := pp.Attributes()
-	rawProfile, ok := attrs["raw_profile"].(SSOCustomTokenClaims)
-	if !ok {
-		return
+func (p *MockProvider) ListPrincipalsByUserID(userID string) ([]principal.Principal, error) {
+	var principals []principal.Principal
+	for _, p := range p.PrincipalMap {
+		if p.UserID == userID {
+			var principal principal.Principal
+			principal = &p
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/dependency/principal/customtoken/mock_provider.go]: What's wrong?

**Comment by louischan-oursky** [comment]
@kiootic Addressed all comments except the loop variable capturing one.

**Comment by louischan-oursky** [comment]
@kiootic The wrong loop variable capturing was also fixed.

File: pkg/auth/handler/sso/sso.go [hunk 3]
```diff
@@ -315,24 +320,40 @@ func (h respHandler) findExistingOAuthPrincipal(oauthAuthInfo sso.AuthInfo) (*oa
...
-		return nil, nil
-	}
+	principals, err := h.IdentityProvider.ListPrincipalsByClaim("email", oauthAuthInfo.ProviderUserInfo.Email)
 	if err != nil {
 		return nil, err
 	}
-	if !h.PasswordAuthProvider.CheckLoginIDKeyType(passwordPrincipal.LoginIDKey, metadata.Email) {
-		return nil, nil
+	var filteredPrincipals []principal.Principal
+	for _, p := range principals {
```
Comment (kiootic) [inline pkg/auth/handler/sso/sso.go]: we can use continue for this loop too.

Comment (louischan-oursky) [reply pkg/auth/handler/sso/sso.go]: @kiootic simplified.

**Pull request merged**